_posts/2019-08-23-til-ecto-select-merge.md

```diff
@@ -0,0 +1,237 @@
+
+## Option 1: Add an Association and Preload the Data
+
+If we associate User and AdmissionEvent, then we can preload the associated User record and read the full name directly from there.
+
+```elixir
+defmodule Registrar.Tracking.AdmissionEvent do
+  use Ecto.Schema
+  import Ecto.Query, only: [from: 2]
+  alias Registrar.Tracking.AdmissionEvent
+  alias Registrar.User
+
+  schema "admission_events" do
+    field(:action, :string)
+    field(:admission_id, :integer)
+    field(:admitter_uuid, :uuid)
+    field(:occurred_at, :naive_datetime)
+
+    # New association
+    belongs_to(:admitter, User, foreign_key: :uuid)
+  end
+end
+
+defmodule Registrar.User do
+  use Ecto.Schema
+  alias Registrar.Tracking.AdmissionEvent
+
+  schema "users" do
+    field(:uuid, :uuid)
+    field(:full_name, :string)
+
+    # New association
+    has_many(:admission_events, AdmissionEvent)
+  end
+end
+```
+
+```elixir
+# Trying out our new association...
+
+iex> admission = Repo.get(Admission, 1)
+iex> events = AdmissionEvent.for_admission(admission) |> Repo.all() |> Repo.preload(:admitter)
+[
+  %Registrar.Tracking.AdmissionEvent{
+    __meta__: #Ecto.Schema.Metadata<:loaded, "admission_events">,
+    action: "Student Admitted",
+    admission_id: 3,
+    admitter: %Registrar.User{
+      __meta__: #Ecto.Schema.Metadata<:loaded, "users">,
+      id: 1,
+      name: "Albus Dumbledore",
+      uuid: "7edd4d7f-a790-41f9-b4ef-16f1dc3b33ea"
+    },
+    admitter_uuid: "7edd4d7f-a790-41f9-b4ef-16f1dc3b33ea",
+    id: 1,
+    occurred_at: ~N[2019-07-29 02:22:18]
+  }
+]
+iex> event = List.first(events)
+iex> event.admitter.name
+"Albus Dumbledore"
+```
+
+This approach gets the job done, but it's a little heavy. We only need the admitter's full name, so why retrieve an entire `User` struct? You can also see how this pattern could lead to a super cluttered User schema. Right now it has many `admission_events`, but soon it could have many `application_events`, `interview_events`, `billing_events`, etc.
+
+## Option 2: ✨ Select Merge ✨
+
+`Ecto.Query#select_merge/3` gives us an option that's much more succinct and precise. Check out this slickness:
+
+```elixir
+defmodule Registrar.Tracking.AdmissionEvent do
+  use Ecto.Schema
+  import Ecto.Query, only: [from: 2]
+  alias Registrar.User
+  alias Registrar.Tracking.AdmissionEvent
+
+  schema "admission_events" do
+    field(:action, :string)
+    field(:admission_id, :integer)
+    field(:admitter_uuid, :uuid)
+    field(:occurred_at, :naive_datetime)
+
+    ###### STEP ONE #######
+    #  Add Virtual Field  #
+    #######################
+    field(:admitter_name, :string, virtual: true)
+  end
+
+  def for_admission(query \\ AdmissionEvent, admission) do
+    from(ae in query,
+      where: ae.admission_id == ^admission.id,
+      order_by: [desc: ae.occurred_at],
+
+      #### STEP TWO ####
+      #  Join on User  #
+      ##################
+      join: u in User,
+      on: r.admitter_uuid == u.uuid,
```

Comment:
where does the `r` variable come from in the query? I couldn't get the code to compile.  I got this error:
```
** (Ecto.Query.CompileError) unbound variable `r` in query. If you are attempting to interpolate a value, use ^var
    lib/ecto/query.ex:755: Ecto.Query.from/5
    expanding macro: Ecto.Query.from/2
    /tracking/admission_event.ex:22: EctoSelectMerge.Tracking.AdmissionEvent.for_admission/2
    (elixir) lib/kernel/parallel_compiler.ex:229: anonymous fn/4 in Kernel.ParallelCompiler.spawn_workers/7
```

Could you please clarify this? :)

Reply:
yep, will fix. it's a copy pasta mistake :)